### PR TITLE
[Routing] Added trim flag to Route::setPath

### PR DESCRIPTION
**Pull request by ragboyjr**
| Q | A |
| --- | --- |
| Branch? | master |
| Bug fix? | no |
| New feature? | yes |
| BC breaks? | yes |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | none |
| License | MIT |
| Doc PR | N/A |

silexphp/silex#149

Added a trim flag to the Route setPath. This replaces #18701 

Signed-off-by: RJ Garcia rj@bighead.net

### CODE REVIEW

**Comment by ragboyjr** [comment]
Hmm, something in the DependencyInjection component is making travis fail on hhvm.

**Comment by ragboyjr** [comment]
@Ener-Getick, @stof, @GromNaN, Oh, you know, I just though of another solution that could actually work for everyone.

What if we added a new protected method in the Route like this

```php
public function setPath($pattern)
{
    // A pattern must start with a slash and must not have multiple slashes at the beginning because the
    // generated path for this route would be confused with a network path, e.g. '//domain.com/path'.
    return $this->doSetPath('/'.ltrim(trim($pattern), '/'));
}

protected function doSetPath($pattern) 
{
    $this->path = $pattern;
    $this->compiled = null;

    return $this;
}
```

or, instead of protected, we could always make `doSetPath` public and change the name to something like `setPathRaw` or whatever.

Either way, this wouldn't break BC, and is a cleaner solution that prevents "monkey patching"

**Comment by xabbuh** [comment]
Can you describe your use case where you would like to have trimming disabled?

**Comment by ragboyjr** [comment]
@xabbuh silexphp/Silex#149

File: src/Symfony/Component/Routing/Route.php [hunk 1]
```diff
@@ -148,14 +148,15 @@ public function getPath()
...
      *
      * @return Route The current Route instance
      */
-    public function setPath($pattern)
+    public function setPath($pattern, $trim = true)
     {
         // A pattern must start with a slash and must not have multiple slashes at the beginning because the
         // generated path for this route would be confused with a network path, e.g. '//domain.com/path'.
-        $this->path = '/'.ltrim(trim($pattern), '/');
+        $this->path = $trim ? '/'.ltrim(trim($pattern), '/') : $pattern;
```
Comment (Tobion) [inline src/Symfony/Component/Routing/Route.php]: the double slash removal must be kept as described in the comment above

Comment (ragboyjr) [inline src/Symfony/Component/Routing/Route.php]: Well, what would you think about doing this instead: https://github.com/symfony/symfony/pull/18702#issuecomment-216617393

**Comment by Ener-Getick** [comment]
It took me some time to understand the issue so in case it helps here is an illustration with a routing file:
```yml
// routing.yml
sub_routing:
   resource: sub_routing.yml
   prefix: /foo
```
```yml
// sub_routing.yml
my_route:
    path: ''

// you would expect to be able to access /foo here 
// but you can only access /foo/ as pathes are always
// prefixed with a /
```

**Comment by ragboyjr** [comment]
@Ener-Getick, that is precisely it. I wasn't accustomed to the Syfmony configuration for this because I was referring to Silex, but, yes it's the same exact issue.

**Comment by Ener-Getick** [comment]
Another solution could be to detect empty paths and to not format them.

**Comment by ragboyjr** [comment]
@Ener-Getick, that would work for me as well, but that would be also be a BC break, unless we considered this a bug in the way it handles empty urls.

**Comment by javiereguiluz** [comment]
This seems related to #12141 too.

**Comment by fabpot** [comment]
I'm closing here as what @GuilhemN describes is exactly how this should work. When mounting a resource, like on a filesystem, the prefix is a directory.

**Comment by ragboyjr** [comment]
@fabpot sounds good, thanks.

**Pull request closed without merging**